### PR TITLE
mastodon: easier build patching

### PR DESCRIPTION
**Pull request by farcaller**
## Description of changes

This allows to apply patches to whatever version of mastodon is in the nixpkgs without owning a local copy of `source.nix`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by mrcjkb — APPROVED** [review]

**Review by cyplo — APPROVED** [review]

**Comment by mrcjkb** [comment]
Result of `nixpkgs-review pr 268551` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

**Comment by erictapen** [comment]
Can't you just use `applyPatches` to build a source from patches?

**Comment by farcaller** [comment]
@erictapen It just feels less ergonomic, given there is already a call to `applyPatches` inside. You can absolutely do this with the current code but it will look something akin to

```
  services.mastodon.package = pkgs.callPackage "${pkgs.path}/pkgs/servers/mastodon" (
    let
      oldSrc = pkgs.callPackage "${pkgs.path}/pkgs/servers/mastodon/source.nix" { };
    in
    {
      # patches = [ ./charlimit.patch ];
      srcOverride = (pkgs.applyPatches {
        src = oldSrc;
        patches = [ ];
      } // { inherit (oldSrc) version yarnHash; });
    }
  );
```

**Review by 999eagle — APPROVED** [review]
I've been patching mastodon using an overlay like this:

```nix
self: super: {
  mastodon-custom = super.mastodon.override {
    srcOverride = super.applyPatches {
      inherit (super.mastodon) src;
      inherit (super.mastodon.src) version yarnHash;
      patches = [ ... ];
    };
  };
```

This PR would make that a bit easier to do. LGTM.

File: pkgs/servers/mastodon/update.sh [hunk 1]
```diff
@@ -94,7 +94,7 @@ in
       rev = "v\${version}";
       hash = "$HASH";
     };
-    patches = [$PATCHES];
+    inherit patches;
```
Comment (erictapen) [inline pkgs/servers/mastodon/update.sh:97]:
```suggestion
    patches = patches ++ [ $PATCHES];
```

The update script is capable of taking patches as arguments. To me this functionality is overkill, but let's not break that for now.

Comment (farcaller) [reply pkgs/servers/mastodon/update.sh]: Thanks for catching that! I just did a blind copy-paste in the script without ever wondering why there was an empty list in there.

**Review by erictapen — APPROVED** [review]
Could you meld the two commits into one or prefix the second commit message with `mastodon: `?

**Comment by farcaller** [comment]
Sorry, screwed up the change in the codespace. Should be good now.

**Review by 999eagle — APPROVED** [review]

**Pull request merged**